### PR TITLE
[Attention][MLA][DCP] Query replication for MLA decode (DeepSeek-V2/R1 + Kimi-K2.5)

### PR DESCRIPTION
**Pull request by sungsooha**
## Purpose

With Decode Context Parallelism (DCP) the KV cache is sharded across the DCP group, so the standard MLA decode path **all-gathers the query** across the group every step so each rank can attend its KV shard with the full head set, then LSE-reduces the partials. That all-gather sits on the decode critical path.

This PR adds an **opt-in** alternative: replicate the (small) MLA query projection *within each DCP group* at load time, so every rank already materializes the full group-local query head set locally. Decode then **skips the query all-gather** and
runs a group-local `W_UK_T_dcp_qrep` BMM. The query projection is tiny relative to the KV cache, so replicating it costs little memory while removing a collective from the decode hot path. Gated behind `VLLM_DCP_Q_REPLICATE` (default off → exact current behavior).

RFC: #34018 (prematurely closed)

## Scope & dependencies

qrep is a **self-contained, opt-in feature**: it replicates the MLA query projection within each DCP group so decode can skip the query all-gather. It is **backend-agnostic** — it only requires the MLA decode backend to return decode-LSE
(a base-DCP property) and does not re-implement LSE or KV handling. On any backend whose DCP decode-LSE is already merged, qrep runs on merged DCP alone.

Validated here with `--attention-backend FLASHINFER_MLA`; that backend's DCP decode-LSE is in flight upstream (#44044 / #43729), so reproducing the evidence below requires it — mentioned for reproducibility, not because qrep is coupled to it.

## Compatibility & roadmap

**Supported today (opt-in, `VLLM_DCP_Q_REPLICATE=1`, DCP world size > 1):**

| Dimension | qrep support | Notes |
|---|---|---|
| MLA decode backend | **FlashInfer-MLA** ✓ | needs the backend's DCP decode-LSE (#44044 / #43729) |
| Head-padding MLA backends (e.g. CutlassMLA `q_pad_num_heads`) | ✗ guarded | `W_UK_T_dcp_qrep` is built unpadded `dcp*num_heads`; validated once at weight-load |
| aiter FP4 / FP8 triton MLA BMM | ✗ guarded | qrep builds the BMM weight on the bf16 path only; validated once at weight-load |
| KV-cache dtype | **fp8 ✓** / bf16 ✓ | orthogonal — qrep replicates the *query projection*, not the KV cache (validated with `--kv-cache-dtype fp8_e4m3`). |
| q / kv projection **weights** | unquantized bf16 ✓ | quantized q/kv sources raise `NotImplementedError` (see roadmap) |
| Models | DeepSeek-V2/V3/R1 (dense MLA), Kimi-K2.5/2.6 (wrapper loader) | |
| Default-off / DCP=1 | exact baseline | pure no-op; every touched site takes the original branch |

**Roadmap (follow-ups, not in this PR):**
1. **DSV4 support** along with DCP support (WIP, Preparing for two PRs specifically for DSV4)
2. **Quantized q/kv projection replicas.**
3. **Head-padding MLA backends.** Would require a padded `W_UK_T_dcp_qrep` to serve CutlassMLA-style `q_pad_num_heads`.
4. **Per-subset (decode-only) projection.** Investigated as a way to drop the projection's prefill overhead; currently shelved because it did not beat the compute-both path in decode-dominant traffic. Revisit only with a graph-friendly form.

## Test Plan

Three matched arms (same image/model/harness), toggling only DCP + qrep, on both models:

| Arm | TP | DCP | DCP backend | VLLM_DCP_Q_REPLICATE |
|---|---|---|---|---|
| DCP (all-gather) | 8 | 8 | `--dcp-comm-backend a2a` | 0 |
| DCP + qrep | 8 | 8 | `--dcp-comm-backend a2a` | 1 |

### Reproduction (serve args / commands)

Enable qrep with `VLLM_DCP_Q_REPLICATE=1` + `--decode-context-parallel-size N>1`
(gated; no-op at DCP=1). No CLI/EngineArg change.

```bash
export VLLM_DCP_Q_REPLICATE=${QREP}   # 0 or 1, per arm
vllm serve ${MODEL} \
  --trust-remote-code --enable-expert-parallel \
  --tensor-parallel-size ${TP} \
  --decode-context-parallel-size ${DCP} \
  --attention-backend FLASHINFER_MLA \
  ${DCP_BACKEND} \
  --max-model-len ${MML} --max-num-seqs ${MNS} \
  --enable-prefix-caching --enable-chunked-prefill
# Kimi only (vision encoder): add
#   --skip-mm-profiling --limit-mm-per-prompt '{"image":0,"video":0}'
```

## Test Result

**Setup.** 
- (8× B200); **TP8 / DCP8**
- `--dcp-comm-backend a2a` 
- `--attention-backend FLASHINFER_MLA` (`mla_prefill_backend=FLASHINFER`) 
- fp8 KV 
- `--max-num-seqs 32`
-  **FULL CUDA graph** (capture `1 2 4 8 16 32`). 
- Two regimes: 
    - **decode-dominant** (random 128/1024)
    - **mixed** (1024/1024)
    - 256 prompts
    - concurrency 32.

**Decode latency / throughput — qrep removes the decode query all-gather:**

| Model | Regime | qrep P50 TPOT (ms) | all-gather P50 TPOT (ms) | TPOT Δ | qrep tput (tok/s) | all-gather tput (tok/s) | tput Δ |
|---|---|---:|---:|---:|---:|---:|---:|
| DeepSeek-R1-NVFP4 | decode | 16.55 | 17.04 | **−2.88%** | 1891.14 | 1837.62 | **+2.91%** |
| DeepSeek-R1-NVFP4 | mixed | 17.57 | 18.07 | **−2.78%** | 1758.09 | 1721.94 | **+2.10%** |
| Kimi-K2.6-NVFP4 | decode | 15.32 | 16.04 | **−4.51%** | 2052.93 | 1959.96 | **+4.74%** |
| Kimi-K2.6-NVFP4 | mixed | 15.97 | 16.60 | **−3.75%** | 1942.95 | 1867.40 | **+4.05%** |

**GSM8K accuracy (full-1319, 5-shot, c32) — qrep is accuracy-equivalent:**

| Model | qrep (flex / strict) | all-gather (flex / strict) | Δ flex (pp) |
|---|---|---|---:|
| DeepSeek-R1-NVFP4 | 0.9553 / 0.9538 | 0.9538 / 0.9530 | +0.15 |
| Kimi-K2.6-NVFP4 | 0.9371 / 0.9363 | 0.9393 / 0.9386 | −0.22 |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/deepseek_v2.py [hunk 1]
```diff
@@ -914,8 +930,33 @@ def __init__(
...
+        )
+        self.dcp_q_b_proj = None
+        self.dcp_q_proj = None
+        self.dcp_kv_b_proj = None
+
+        def build_dcp_qrep_replica(
+            source: nn.Module, in_features: int, per_head_out: int, prefix: str
+        ) -> ReplicatedLinear:
+            # qrep replica: unquantized, loads the checkpoint's bf16 DCP-group rows.
+            _require_unquantized_dcp_qrep_source(source, prefix)
```
Comment (LucasWilkinson) [inline vllm/model_executor/models/deepseek_v2.py]: why is this required?

**Review by LucasWilkinson — COMMENTED** [review]
Apologies for the long review delay! can you please rebase/merge latest main?

Thank you for contributing this!

File: vllm/model_executor/models/deepseek_v2.py [hunk 2]
```diff
@@ -914,8 +930,33 @@ def __init__(
...
+        self.dcp_q_replicate = (
+            dcp_q_replicate_enabled()
+            and self.dcp_world_size > 1
+            and getattr(vllm_config.model_config, "use_mla", True)
+        )
+        self.dcp_q_b_proj = None
+        self.dcp_q_proj = None
+        self.dcp_kv_b_proj = None
+
+        def build_dcp_qrep_replica(
```
Comment (LucasWilkinson) [inline vllm/model_executor/models/deepseek_v2.py]: can we please inline this?

**Review by LucasWilkinson — COMMENTED** [review]
do you think it would be possible to reduce the number of branches in vllm/model_executor/layers/attention/mla_attention.py ? example for kv_b_proj could we maybe not have dcp_kv_b_proj but just pass kv_b_proj as replicated?

**Comment by sungsooha** [comment]
> Apologies for the long review delay! can you please rebase/merge latest main?
> 
> Thank you for contributing this!

Rebased.

**Comment by sungsooha** [comment]
> do you think it would be possible to reduce the number of branches in vllm/model_executor/layers/attention/mla_attention.py ? example for kv_b_proj could we maybe not have dcp_kv_b_proj but just pass kv_b_proj as replicated?

Done. kv_b_proj is now built as the replicated projection directly when qrep is on (no more parallel dcp_kv_b_proj), and the local-head slice is taken via a single _kv_b_proj_local_heads helper that also de-duplicates the three prefill call sites. The decode block simplifies to "cat if still a tuple; skip the all-gather when qrep is on."

I did keep q_b_proj/q_proj with separate dcp_q_* replicas rather than unifying them in place, because prefill needs only the local query while decode needs the full group-replicated one — computing both (cheap local + replicated) beat compute-full-then-slice in decode-dominant traffic. Happy to unify it the same way if you'd prefer.

File: vllm/model_executor/models/deepseek_v2.py (hunk 1 above)
Comment (sungsooha) [reply vllm/model_executor/models/deepseek_v2.py]: This was the unquantized-source guard for the parallel dcp_kv_b_proj. That module is gone now (kv_b_proj itself is the replicated projection), so the line is removed.

File: vllm/model_executor/models/deepseek_v2.py (hunk 2 above)
Comment (sungsooha) [reply vllm/model_executor/models/deepseek_v2.py]: Inlined

**Comment by sungsooha** [comment]
@LucasWilkinson  After refactor.

DSR1 TP8/DCP8 FI-MLA fp8-KV a2a, **GB200**, FULL cudagraph
(decode,FULL sizes 1-32), ISL128/OSL1024 c32, 256 prompts, 0 errors. **Perf A/B — reproduces published (+2.9%):**

| Metric | OFF (all-gather) | ON (qrep) | Δ |
|---|--:|--:|--:|
| Throughput tok/s | 1386.57 | 1425.75 | **+2.83%** |
| P50 TPOT ms | 22.201 | 21.859 | −1.54% |
| tps-user p50 | 45.04 | 45.75 | +1.57% |
| TTFT p50 ms | 644.83 | 642.47 | −0.37% |

**GSM8K parity (full 1319, 5-shot-equiv, c32) — accuracy-equivalent (matches published +0.15pp):**
| Arm | accuracy |
|---|---|
| OFF | 0.9469 |
| ON | 0.9485 (**+0.15pp**, noise — qrep is exact) |

**Comment by LucasWilkinson** [comment]
@sungsooha what about something like https://github.com/sungsooha/vllm/pull/1 i.e. make a custom `DCPGroupColumnParallelLinear` `ColumnParallelLinear` subclass; I think this can make this alot less invasive

(this PR is vibecoded and mostly for demonstration purposes; would need to be de-slopified)

**Comment by sungsooha** [comment]
@LucasWilkinson 

Thanks for sharing the `DCPGroupColumnParallelLinear` implementation — much cleaner shape. I de-slopified it and pushed it onto the branch (kept it as your commit). It drops the `dcp_q_*` replica modules + custom weight loading, the `kv_b_proj` special-casing, and the Kimi-K2.5 mirroring; qrep is now `forward()` (drop-in `ColumnParallelLinear`) vs `forward_replicated()`, with standard/quantized loading unchanged.

Re-validated on the new machinery (GB200, DSR1-NVFP4, TP8 / DCP8, fp8-KV, FULL cudagraph). Serve (2-node `mp`, node-0 shown; `VLLM_DCP_Q_REPLICATE` toggles off/on):

```bash
VLLM_DCP_Q_REPLICATE=0 vllm serve nvidia/DeepSeek-R1-0528-NVFP4-v2 --trust-remote-code \
  --tensor-parallel-size 8 --decode-context-parallel-size 8 --enable-expert-parallel \
  --distributed-executor-backend mp --nnodes 2 --node-rank 0 --master-addr $MASTER --master-port 29842 \
  --max-model-len 8192 --max-num-seqs 64 --max-num-batched-tokens 8192 --gpu-memory-utilization 0.90 \
  --kv-cache-dtype fp8_e4m3 --attention-backend FLASHINFER_MLA --dcp-comm-backend a2a \
  --enable-chunked-prefill -cc.cudagraph_mode=FULL --host 0.0.0.0 --port 8000
```

### GSM8K (1319, 5-shot) — accuracy-equivalent
| arm | accuracy | correct |
|---|---|---|
| off | 0.9507 | 1254 / 1319 |
| on  | 0.9469 | 1249 / 1319 |

Δ −0.38 pp (5 questions), 0 request errors.

### Decode benchmark (AIPerf) — qrep-on +7.9% tok/s
Chat/streaming, concurrency 32, synthetic ISL 128 / OSL 1024, 64 requests (`min_tokens=OSL`, `ignore_eos`, unique prefix per request):

```bash
aiperf profile -m dsr1-qrep-gate --tokenizer nvidia/DeepSeek-R1-0528-NVFP4-v2 \
  --url http://$MASTER:8000 --endpoint-type chat --streaming \
  --concurrency 32 --synthetic-input-tokens-mean 128 --output-tokens-mean 1024 \
  --extra-inputs min_tokens:1024 --extra-inputs ignore_eos:true --request-count 64
```

| arm | tok/s | ITL p50 |
|---|---|---|
| off | 1613.7 | 19.37 ms |
| on  | 1741.2 | 18.03 ms |

Default-off is bit-identical (also checked on DSV2-Lite bf16).

### Sparse MLA (DSA)
Also validated on GLM-5.2 (DSA): qrep-on produces identical output and is faster, so qrep applies uniformly to dense and sparse MLA. Serve (4-node `mp`, node-0):

```bash
VLLM_DCP_Q_REPLICATE=0 vllm serve nvidia/GLM-5.2-NVFP4 --trust-remote-code \
  --tensor-parallel-size 16 --decode-context-parallel-size 16 --enable-expert-parallel \
  --distributed-executor-backend mp --nnodes 4 --node-rank 0 --master-addr $MASTER --master-port 29742 \
  --max-model-len 8192 --max-num-seqs 512 --max-num-batched-tokens 8192 --gpu-memory-utilization 0.90 \
  --kv-cache-dtype fp8_e4m3 --attention-backend FLASHINFER_MLA_SPARSE --dcp-comm-backend a2a \
  --cp-kv-cache-interleave-size 1 --block-size 64 --enable-chunked-prefill \
  -cc.cudagraph_mode=FULL --host 0.0.0.0 --port 8000
```

Ready for another look.

**Review by LucasWilkinson — APPROVED** [review]
Overall looks good to me but I think @MatthewBonanni should take a look

File: vllm/model_executor/layers/linear.py [hunk 3]
```diff
@@ -577,6 +607,112 @@ def extra_repr(self) -> str:
...
+            else self._local_weight
+        )
+        assert weight is not None, (
+            "build_local_weight() must run before forward() on a quantized "
+            f"{type(self).__name__} ({self.prefix})"
+        )
+        bias = self.bias[self._local_shard] if self.bias is not None else None
+        output = torch.nn.functional.linear(
+            input_, weight, None if self.skip_bias_add else bias
+        )
```
Comment (MatthewBonanni) [inline vllm/model_executor/layers/linear.py]: So does this not support quantized gemm? That seems like a regression when `VLLM_DCP_Q_REPLICATE` is off

File: vllm/model_executor/layers/linear.py [hunk 4]
```diff
@@ -577,6 +607,112 @@ def extra_repr(self) -> str:
...
+    def __init__(self, *args, **kwargs):
+        dcp_world_size = (
+            get_current_vllm_config().parallel_config.decode_context_parallel_size
+        )
+        rank = get_tensor_model_parallel_rank()
+        world_size = get_tensor_model_parallel_world_size()
+        self.group_size = max(dcp_world_size, 1)
+        self.rank_in_group = rank % self.group_size
+        # qrep only decides whether callers use forward_replicated(); the weight
+        # is sharded per DCP group either way.
```
Comment (MatthewBonanni) [inline vllm/model_executor/layers/linear.py]: Don't we want to preserve the old behavior when `VLLM_DCP_Q_REPLICATE` is off? This now replicates the weights by a factor of `dcp_world_size` regardless

File: vllm/model_executor/layers/linear.py (hunk 3 above)
Comment (sungsooha) [reply vllm/model_executor/layers/linear.py]: Right — with qrep off it's now a plain ColumnParallelLinear, so quantized gemm runs unchanged (no build_local_weight / dequantized forward). The dequantized-weight path is taken only when qrep is explicitly on.

File: vllm/model_executor/layers/linear.py (hunk 4 above)
Comment (sungsooha) [reply vllm/model_executor/layers/linear.py]: Fixed — gated `group_size` on `qrep_active`, so with VLLM_DCP_Q_REPLICATE off the layer collapses to group_size == 1, i.e. a plain ColumnParallelLinear (per-rank shard, no replication). The ×`dcp_world_size` weight replication is now paid only when qrep is enabled.

**Comment by MatthewBonanni** [comment]
You can eliminate the unnecessary `W_UV` replication by paying a single `W_UK` allgather during model loading. This also dramatically simplifies the PR.  I implemented this as part of https://github.com/vllm-project/vllm/pull/45964/commits/a661e9c81d731320c2ef970f98b72dc84cfe0f28

I also moved the feature switch to what I think is a more sensible location -- we conditionally choose between `DCPGroupColumnParallelLinear` and `ColumnParallelLinear` rather than making `DCPGroupColumnParallelLinear` behave like `ColumnParallelLinear`

**Review by MatthewBonanni — APPROVED** [review]
